src/main/java/org/roaringbitmap/BitSetUtil.java

```diff
@@ -1,138 +1,120 @@
-			// if it is a tiny bitSet, we can directly build a arrayContainer, without an extra bitSet#toLongArray
-			final RoaringBitmap bitmap = new RoaringBitmap();
-			bitmap.highLowContainer.append((short)0, arrayContainerOf(cardinality, bitSet));
-			return bitmap;
-		} else {
-			// before jdk 1.7
-			return bitmapOf(cardinality, toLongArray(cardinality, bitSet));
-			// starting with jdk.17
-			//return bitmapOf(cardinality, bitSet.toLongArray());
-		}
-	}
-	
-	private static long[] toLongArray(final int cardinality, final BitSet bitSet) {		
-		final long[] words = new long[(bitSet.length() + 63) / 64];
-	    for (int i = bitSet.nextSetBit(0); i >= 0; i = bitSet.nextSetBit(i+1)) {
-            if (i == Integer.MAX_VALUE) {
-                break;
+    /**
+     * Generate a RoaringBitmap out of a long[], each long using little-endian
+     * representation of its bits
+     * 
+     * @see BitSet#toLongArray() for an equivalent
+     * @param words array of longs  (will not be modified)
+     * @return roaring bitmap
+     */
+    public static RoaringBitmap bitmapOf(final long[] words) {
+        // split long[] into blocks.
+        // each block becomes a single container, if any bit is set
+        final RoaringBitmap ans = new RoaringBitmap();
+        int containerIndex = 0;
+        for (int from = 0; from < words.length; from += BLOCK_LENGTH) {
+            final int to = Math.min(from + BLOCK_LENGTH, words.length);
+            final int blockCardinality = cardinality(from, to, words);
+            if (blockCardinality > 0) {
+                ans.highLowContainer.insertNewKeyValueAt(containerIndex++, Util
+                        .highbits(from * Long.SIZE), BitSetUtil.containerOf(
+                        from, to, blockCardinality, words));
             }
-            words[i / 64] |= (1L << i % 64); 
-	    }
-	    return words;
-   }
+        }
+        return ans;
+    }
 
-	/**
-	 * Generate RoaringBitmap out of a long[], each long using little-endian representation of its bits 
-	 * 
-	 * @see BitSet#toLongArray() for an equivalent
-	 * @param bitset
-	 * @return roaring bitmap
-	 */
-    static RoaringBitmap bitmapOf(final long[] words) {
-		final int overallCardinality = cardinality(0, words.length, words);
-		if (overallCardinality == 0) {
-			return new RoaringBitmap();
-		} else {
-			return bitmapOf(overallCardinality, words);
-		}
+    private static Container containerOf(final int from, final int to,
+            final int blockCardinality, final long[] words) {
+        // find the best container available
+        if (blockCardinality <= ArrayContainer.DEFAULT_MAX_SIZE) {
+            // containers with DEFAULT_MAX_SIZE or less integers should be
+            // ArrayContainers
+            return arrayContainerOf(from, to, blockCardinality, words);
+        } else {
+            // otherwise use bitmap container
+            return new BitmapContainer(Arrays.copyOfRange(words, from, from + BLOCK_LENGTH),
```

Comment:
Nah. I have just committed new code that checks for that so that this mistake won't be made. This is really a hard expectation in the code and it is very likely that the bitmaps thus generated would have crashed the code later on.

Now, if you are interested, we could look at a different type of bitmap container that has a range of valid values. That could be interesting... but it is more of a research project at this point.